NDTensors/src/dense/tensoralgebra/contract.jl

```diff
@@ -393,15 +397,30 @@ function _contract!(
 
   #tC = similar(CM)
   #_gemm!(tA, tB, El(α), AM, BM, El(β), CM)
-  @strided mul!(CM, AM, BM, El(α), El(β))
+  CM = mul!!(
+    leaf_parenttype(CM), CM, leaf_parenttype(AM), AM, leaf_parenttype(BM), BM, El(α), El(β)
+  )
 
   if props.permuteC
     Cr = reshape(CM, props.newCrange)
     # TODO: use invperm(pC) here?
-    #@timeit_debug timer "_contract!: permutedims C" begin
-    @strided CT .= permutedims(Cr, props.PC)
+    #@timeit_debug timer "_contract!: permutedims C" begin 
+    CT .= permutedims!!(leaf_parenttype(Cr), Cr, props.PC)
```

Comment:
Like my above comment, I think this should be `permutedims!!(leaf_parenttype(CT), CT, leaf_parenttype(Cr), Cr, props.PC)`.

There are a few other places you use `permutedims!!` where similar changes should be made.